cmd/root.go

```diff
@@ -135,13 +139,31 @@ func startServer(system *core.System) error {
 		r.Routes(echoServer)
 	}
 
-	defer func() {
-		if err := system.Shutdown(); err != nil {
-			logrus.Error("Error shutting down system:", err)
-		}
+	// Make sure shutdown is called just once, because we also have it called deferred
+	shutdownOnce := sync.Once{}
+	shutdown := func() {
+		shutdownOnce.Do(func() {
+			logrus.Info("Shutting down...")
+			echoServer.Shutdown()
+			if err := system.Shutdown(); err != nil {
+				logrus.Error("Error shutting down system:", err)
+			}
+			logrus.Info("Shutdown complete. Goodbye!")
+		})
+	}
+	defer shutdown()
```

Comment:
I think it's cleaner to create a new cancellable context with the original context as parent. When the parent context is cancelled it will be propagated to the child and otherwise we will cancel the child context ourselves anyway. Then we can also remove `defer shutdown()` and the `sync.Once` stuff.